src/output-patcher.js

```diff
@@ -0,0 +1,80 @@
+import FSTree from 'fs-tree-diff';
+import md5Hex from 'md5-hex';
+import * as fs from 'fs';
+
+export default class OutputPatcher {
+  constructor(outputPath, logger) {
+    this.outputPath = outputPath;
+    this.entries = [];
+    this.contents = Object.create(null);
+    this.lastTree = FSTree.fromEntries([]);
+    this.isUnchanged = (entryA, entryB) => {
+      if (entryA.isDirectory() && entryB.isDirectory()) {
+        return true;
+      }
+      if (entryA.mode === entryB.mode && entryA.checksum === entryB.checksum) {
+        logger.debug('cache hit, no change to: %s', entryA.relativePath);
+        return true;
+      }
+      logger.debug('cache miss, write to: %s', entryA.relativePath);
+      return false;
+    }
+  }
+
+  // relativePath should be without leading '/' and use forward slashes
+  add(relativePath, content) {
+    this.entries.push(new Entry(relativePath, md5Hex(content)));
+    this.contents[relativePath] = content;
+  }
+
+  patch() {
+    try {
+      this.lastTree = this._patch();
+    } catch (e) {
+      // next build just output everything
+      this.lastTree = FSTree.fromEntries([]);
```

Comment:
needs rebuild state from walk-sync otherwise patching will fail trying to mkdir existing directories, or use fs extra and still call mkdirpSync